### ISSUE TITLE
Unit Testing problem related to reseting

### ISSUE THREAD

**Issue by andreigabreanu** [mobxjs/mobx#227]
I've been searching for a few hours and probably I am missing something but I have an issue with unit testing a react component connected to an MobX enhanced derived store (based on 2 others computed/observable stores).

Here is a snippet of code (using karma + jasmine + webpack + ts-loader)

```js
describe("SomeComponent", () => {
    beforeEach(() => {
       this.store1 = new Store1();
       this.store2 = new Store2();
      this.derivedStore = new DerivedStore(this.store1, this.store2);
    });

   it('should be able to render', () => {
      const $comp = $(<SomeComponent computedStore={this.computedStore} />);
      // assert that $comp.find('.someClass') === 'some text' => OK
   });

  it('should be able to render while doing async', (done) => {
      const $comp = $(<SomeComponent computedStore={this.computedStore} />);
      this.store1.doSomeAsync().then( () => done() );  
       // assert that $comp.find('.someClass') === 'some other text' => STILL OK
   });

  it('should be able to render while doing async #2', (done) => {
      const $comp = $(<SomeComponent computedStore={this.computedStore} />);
      this.store1.doSomeAsync().then( () => done() );  
       // assert that $comp.find('.someClass') === 'some other text' => FAILS 
   });

```

The failure gives the following error(s):

ERROR: '[mobx] An uncaught exception occurred while calculating your computed value, autorun or transformer. Or inside the render method of a React component. These methods should never throw exceptions as MobX will usually not be able to recover from them. Please enable 'Pause on (caught) exceptions' in your debugger to find the root cause. In: 'DerivedStore@1433 / Prop "isLoading"#1434''

Error: [ComputedValue] It is not allowed to assign new values to computed properties. 

If i remove the "throwingComputedValueSetter" from mobx.js and replace it with a () => null everything works OK. 
Also, in the browser everything is good. 
FYI, the stores / components are not doing anything weird, mostly its based on the docs - using the decorators everywhere (@observable, @observer, @computed )
Also, if I keep the React Component renderer only in the first test, and just run the async methods in the 2nd, 3rd + tests, it works fine. So whenever I mount the component more than once, on the next test when something happens in the state it fails. I did try to run unmount() after each test but to no avail.

My investigation leads me to believe that, even though beforeEach is run and I am creating a new instance of the stores, mobx somehow still keeps track of older stuff and ... something happens.

Maybe it is something very obvious that I am missing but do you have any ideas of what could be wrong?

**Comment by mweststrate** [comment]
Note that you seem to mix up  `derivedStore` and `computedStore`.

Besides that I recommend doing `mobx.extras.resetGlobalState()` in the beforeEach. By default MobX stops running after the first exception for fail fast reasons. This function reset all derivations for a fresh. I'll make sure a better error message is given when MobX goes in 'errored' state.

**Comment by andreigabreanu** [comment]
Indeed, I think the store that is using the other two is a "computedStore" - since it does nothing but observe the other 2 and provide computed values for subscribers. Would you say that is correct?

Regarding the error: I tried with the mobx.extras.resetGlobalState() but got the same issue. I did some more investigations last night and it seems the error is not triggered if I don't render any component. I assume that's normal / expected since if no component is mounted, there is no observer thus mobx is not recomputing anything.

My solution right now is to comment the "Throw error" line in mobx and everything works fine, both in the unit tests and in the browser - although I feel uncomfortable since the Error is there for a reason. 

Basically, my goal is to achieve the reactions in the browser via mobx when something is changing, but in my tests, I prefer to run everything from scratch in each test (aka rerender the component whenever I know something has changed and test the result).

Would it help if I attached the component / stores code ?

**Comment by mweststrate** [comment]
Hmm wait I think I didn't read your error well enough earlier: `[ComputedValue] It is not allowed to assign new values to computed properties.`, I guess you are, as suggested by the error, try to assign a new value to something decorated with `@computed`. If that isn't the case, please attach some more code then I'll take a closer look.

**Comment by mweststrate** [comment]
@andreigabreanu could you indicate whether your issue has been resolved? Tnx!

**Comment by andreigabreanu** [comment]
Unfortunately, I didn't have any time to get back to this. I guess we can close for now and I'll reopen when I have more info / time to dig into it. Thanks for the help!
